### ISSUE TITLE
Unable to Add YubiKey

### ISSUE THREAD

**Issue by Sylver-Dragon** [nextcloud/twofactor_u2f#916]
### Steps to reproduce
1. Open **User** -> **Settings** -> **Security**
2. Click **Add U2F Device**

### Expected behaviour
User is prompted to insert/touch YubiKey.

### Actual behaviour
Waiting spinner is seen, user is never prompted for device activation.

### Server configuration

**Operating system**: Nextcloud Docker Container using `latest` tag.  

**Web server:** Builtin Apache web server inside container with NGinx reverse proxy providing TLS.

**Database:** SQLite 3

**PHP version:** 7.4.21

**Version:** (see admin page) 22.0.0

**Updated from an older version or fresh install:** Updated from version 20

**List of activated apps:**
```
Enabled:
  - accessibility: 1.7.0
  - activity: 2.15.0
  - admin_audit: 1.11.0
  - bruteforcesettings: 2.2.0
  - circles: 22.0.0
  - cloud_federation_api: 1.4.0
  - contactsinteraction: 1.2.0
  - dav: 1.18.0
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_pdfviewer: 2.3.0
  - files_rightclick: 1.1.0
  - files_sharing: 1.13.2
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.11.0
  - logreader: 2.7.0
  - lookup_server_connector: 1.9.0
  - nextcloud_announcements: 1.11.0
  - notifications: 2.10.1
  - oauth2: 1.9.0
  - password_policy: 1.12.0
  - photos: 1.4.0
  - privacy: 1.6.0
  - provisioning_api: 1.11.0
  - recommendations: 1.1.0
  - serverinfo: 1.12.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - support: 1.5.0
  - survey_client: 1.10.0
  - systemtags: 1.11.0
  - text: 3.3.0
  - theming: 1.12.0
  - twofactor_backupcodes: 1.10.1
  - twofactor_u2f: 6.2.0
  - viewer: 1.6.0
  - workflowengine: 2.3.0
Disabled:
  - comments
  - dashboard
  - encryption
  - files_external
  - firstrunwizard
  - updatenotification
  - user_ldap
  - user_status
  - weather_status
```

**The content of config/config.php:**
```
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.dragon-bytes.com",
            "stor"
        ],
        "overwwrite.cli.url": "https:\/\/cloud.dragon-bytes.com",
        "overwritehost": "cloud.dragon-bytes.com",
        "overwriteprotocol": "https",
        "overwritecondaddr": "^10\\.0\\.1\\.4$",
        "forwarded_for_headers": [
            "X-Forwarded-For"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "sqlite3",
        "version": "22.0.0.11",
        "overwrite.cli.url": "http:\/\/stor:8081",
        "installed": true,
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "loglevel": 2,
        "log.condition": {
            "apps": [
                "admin_audit"
            ]
        },
        "maintenance": false,
        "theme": ""
    },
    "apps": {
        "accessibility": {
            "installed_version": "1.7.0",
            "types": "",
            "enabled": "yes"
        },
        "activity": {
            "installed_version": "2.15.0",
            "types": "filesystem",
            "enabled": "yes",
            "enable_email": "no"
        },
        "admin_audit": {
            "installed_version": "1.11.0",
            "enabled": "yes",
            "types": "logging",
            "logfile": "\/var\/log\/nextcloud\/audit.log"
        },
        "backgroundjob": {
            "lastjob": "663"
        },
        "bruteforcesettings": {
            "installed_version": "2.2.0",
            "enabled": "yes",
            "types": ""
        },
        "circles": {
            "installed_version": "22.0.0",
            "types": "filesystem,dav",
            "enabled": "yes",
            "loopback_tmp_scheme": "http"
        },
        "cloud_federation_api": {
            "installed_version": "1.4.0",
            "types": "filesystem",
            "enabled": "yes"
        },
        "comments": {
            "installed_version": "1.10.0",
            "types": "logging",
            "enabled": "no"
        },
        "contactsinteraction": {
            "installed_version": "1.2.0",
            "types": "dav",
            "enabled": "yes"
        },
        "core": {
            "installedat": "1564319643.8267",
            "lastupdatedat": "0",
            "vendor": "nextcloud",
            "public_files": "files_sharing\/public.php",
            "public_webdav": "dav\/appinfo\/v1\/publicwebdav.php",
            "installed.bundles": "[\"CoreBundle\"]",
            "scss.variables": "bd0196fb58700c89ef04726036abac38",
            "lastcron": "1626797642",
            "lastupdateResult": "[]",
            "shareapi_allow_public_upload": "no",
            "shareapi_allow_resharing": "no",
            "theming.variables": "a0c8b1f612e7fcdc3cd045a92bc9067c",
            "enterpriseLogoChecked": "yes",
            "moveavatarsdone": "yes",
            "previewsCleanedUp": "1",
            "oc.integritycheck.checker": "[]"
        },
        "dashboard": {
            "installed_version": "7.0.0",
            "types": "",
            "enabled": "no"
        },
        "dav": {
            "installed_version": "1.18.0",
            "types": "filesystem",
            "enabled": "yes",
            "generateBirthdayCalendar": "no",
            "sendInvitations": "no",
            "regeneratedBirthdayCalendarsForYearFix": "yes",
            "buildCalendarSearchIndex": "yes",
            "buildCalendarReminderIndex": "yes",
            "chunks_migrated": "1",
            "sendEventReminders": "no"
        },
        "federatedfilesharing": {
            "installed_version": "1.11.0",
            "types": "",
            "enabled": "yes"
        },
        "federation": {
            "installed_version": "1.11.0",
            "types": "authentication",
            "enabled": "yes"
        },
        "files": {
            "installed_version": "1.16.0",
            "types": "filesystem",
            "enabled": "yes",
            "cronjob_scan_files": "500"
        },
        "files_external": {
            "user_certificate_scan": "not-run-yet"
        },
        "files_pdfviewer": {
            "installed_version": "2.3.0",
            "types": "",
            "enabled": "yes"
        },
        "files_rightclick": {
            "installed_version": "1.1.0",
            "types": "",
            "enabled": "yes"
        },
        "files_sharing": {
            "installed_version": "1.13.2",
            "types": "filesystem",
            "enabled": "yes",
            "outgoing_server2server_share_enabled": "no",
            "incoming_server2server_share_enabled": "no",
            "lookupServerUploadEnabled": "no"
        },
        "files_texteditor": {
            "installed_version": "2.8.0",
            "types": "",
            "enabled": "no"
        },
        "files_trashbin": {
            "installed_version": "1.11.0",
            "types": "filesystem,dav",
            "enabled": "yes"
        },
        "files_versions": {
            "installed_version": "1.14.0",
            "types": "filesystem,dav",
            "enabled": "yes"
        },
        "files_videoplayer": {
            "installed_version": "1.11.0",
            "types": "",
            "enabled": "yes"
        },
        "firstrunwizard": {
            "installed_version": "2.9.0",
            "types": "logging",
            "enabled": "no"
        },
        "gallery": {
            "installed_version": "18.4.0",
            "types": "",
            "enabled": "no"
        },
        "logreader": {
            "installed_version": "2.7.0",
            "types": "",
            "enabled": "yes"
        },
        "lookup_server_connector": {
            "installed_version": "1.9.0",
            "types": "authentication",
            "enabled": "yes"
        },
        "nextcloud_announcements": {
            "installed_version": "1.11.0",
            "types": "logging",
            "enabled": "yes",
            "pub_date": "Thu, 24 Oct 2019 00:00:00 +0200"
        },
        "notifications": {
            "installed_version": "2.10.1",
            "types": "logging",
            "enabled": "yes"
        },
        "oauth2": {
            "installed_version": "1.9.0",
            "types": "authentication",
            "enabled": "yes"
        },
        "password_policy": {
            "installed_version": "1.12.0",
            "types": "authentication",
            "enabled": "yes",
            "enforceHaveIBeenPwned": "1"
        },
        "photos": {
            "installed_version": "1.4.0",
            "types": "",
            "enabled": "yes"
        },
        "privacy": {
            "installed_version": "1.6.0",
            "types": "",
            "enabled": "yes"
        },
        "provisioning_api": {
            "installed_version": "1.11.0",
            "types": "prevent_group_restriction",
            "enabled": "yes"
        },
        "recommendations": {
            "installed_version": "1.1.0",
            "types": "",
            "enabled": "yes"
        },
        "serverinfo": {
            "installed_version": "1.12.0",
            "types": "",
            "enabled": "yes"
        },
        "settings": {
            "installed_version": "1.3.0",
            "types": "",
            "enabled": "yes"
        },
        "sharebymail": {
            "installed_version": "1.11.0",
            "types": "filesystem",
            "enabled": "yes"
        },
        "support": {
            "installed_version": "1.5.0",
            "types": "session",
            "enabled": "yes",
            "SwitchUpdaterServerHasRun": "yes"
        },
        "survey_client": {
            "installed_version": "1.10.0",
            "types": "",
            "enabled": "yes"
        },
        "systemtags": {
            "installed_version": "1.11.0",
            "types": "logging",
            "enabled": "yes"
        },
        "text": {
            "installed_version": "3.3.0",
            "types": "dav",
            "enabled": "yes"
        },
        "theming": {
            "installed_version": "1.12.0",
            "types": "logging",
            "enabled": "yes",
            "name": "Dragon-Bytes",
            "cachebuster": "9",
            "url": "***REMOVED SENSITIVE VALUE***",
            "slogan": "***REMOVED SENSITIVE VALUE***",
            "color": "#8F3BC9"
        },
        "twofactor_backupcodes": {
            "installed_version": "1.10.1",
            "types": "",
            "enabled": "yes"
        },
        "twofactor_u2f": {
            "installed_version": "6.2.0",
            "enabled": "yes",
            "types": ""
        },
        "updatenotification": {
            "installed_version": "1.10.0",
            "types": "",
            "enabled": "no",
            "update_check_errors": "0",
            "core": "17.0.2.1",
            "twofactor_u2f": "5.0.1"
        },
        "user_status": {
            "installed_version": "1.0.0",
            "types": "",
            "enabled": "no"
        },
        "viewer": {
            "installed_version": "1.6.0",
            "types": "",
            "enabled": "yes"
        },
        "weather_status": {
            "installed_version": "1.0.0",
            "types": "",
            "enabled": "no"
        },
        "workflowengine": {
            "installed_version": "2.3.0",
            "types": "filesystem",
            "enabled": "yes"
        }
    }
}
```
#### Client configuration
**Browser:** FireFox 90

**Operating system:** Windows 10

##### Web server error log
```
N/A
```

##### Server log (data/nextcloud.log)
```
{"reqId":"4F1JoEppB02YrTCaCEbL","level":3,"time":"2021-07-20T15:05:34+00:00","remoteAddr":"10.0.1.3","user":"john","app":"index","method":"POST","url":"/apps/twofactor_u2f/settings/finishregister","message":"Argument 1 passed to OCA\\TwoFactorU2F\\Controller\\SettingsController::finishRegister() must be of the type string, null given, called in /var/www/html/lib/private/AppFramework/Http/Dispatcher.php on line 217","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:90.0) Gecko/20100101 Firefox/90.0","version":"22.0.0.11","exception":{"Exception":"Exception","Message":"Argument 1 passed to OCA\\TwoFactorU2F\\Controller\\SettingsController::finishRegister() must be of the type string, null given, called in /var/www/html/lib/private/AppFramework/Http/Dispatcher.php on line 217","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/App.php","line":156,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorU2F\\Controller\\SettingsController"},"finishRegister"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\TwoFactorU2F\\Controller\\SettingsController","finishRegister",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"twofactor_u2f.settings.finishRegister"}]},{"file":"/var/www/html/lib/base.php","line":1000,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/twofactor_u2f/settings/finishregister"]},{"file":"/var/www/html/index.php","line":36,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","Line":158,"Previous":{"Exception":"TypeError","Message":"Argument 1 passed to OCA\\TwoFactorU2F\\Controller\\SettingsController::finishRegister() must be of the type string, null given, called in /var/www/html/lib/private/AppFramework/Http/Dispatcher.php on line 217","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":217,"function":"finishRegister","class":"OCA\\TwoFactorU2F\\Controller\\SettingsController","type":"->","args":[null,null,"YubiKey"]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":126,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorU2F\\Controller\\SettingsController"},"finishRegister"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":156,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorU2F\\Controller\\SettingsController"},"finishRegister"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\TwoFactorU2F\\Controller\\SettingsController","finishRegister",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"twofactor_u2f.settings.finishRegister"}]},{"file":"/var/www/html/lib/base.php","line":1000,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/twofactor_u2f/settings/finishregister"]},{"file":"/var/www/html/index.php","line":36,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/custom_apps/twofactor_u2f/lib/Controller/SettingsController.php","Line":65},"CustomMessage":"--"}}

```

##### Browser log
```
[DEBUG] twofactor_u2f: starting registration 
Object { app: "twofactor_u2f", uid: "john", appId: "http://cloud.dragon-bytes.com" }
ConsoleLogger.js:29:16


[WARN] twofactor_u2f: Server generated a U2F App ID that starts with HTTP but you're connected via HTTPS. Set your overwrites https://docs.nextcloud.com/server/stable/admin_manual/config…server/reverse_proxy_configuration.html#overwrite-parameters 
Object { app: "twofactor_u2f", uid: "john" }
ConsoleLogger.js:37:16

Uncaught (in promise) Error: U2F device registration failed (error code unknown)
    rejectRegistration AddDeviceDialog.vue:157
    register AddDeviceDialog.vue:135
    promise callback*register AddDeviceDialog.vue:135
    promise callback*start AddDeviceDialog.vue:105
    VueJS 29
        Qtn_wrapperyraexrgpdpuo_updater 
        get fnmount $mount 
        init ppdpuo_updater 
        get fnmount $mount
    <anonymous> main-settings.js:52
    Webpack 3
        n
        <anonymous>
        <anonymous>
AddDeviceDialog.vue:157
```

##### Network Capture  
I used tcpdump to capture the traffic on the container interface.  Following the applicable HTTP stream in WireShark shows:  
```
POST /apps/twofactor_u2f/settings/startregister HTTP/1.0
Host: cloud.dragon-bytes.com
X-Real-IP: 192.168.1.145
X-Forwarded-For: 192.168.1.145
X-Forwarded-Proto: https
Connection: close
Content-Length: 0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:90.0) Gecko/20100101 Firefox/90.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
requesttoken: REDACTED
Origin: https://cloud.dragon-bytes.com
Cookie: __Host-nc_sameSiteCookielax=true; __Host-nc_sameSiteCookiestrict=true; nc_sameSiteCookielax=true; nc_sameSiteCookiestrict=true; nc_username=john; nc_token=REDACTED; nc_session_id=REDACTED; oc_sessionPassphrase=REDACTED; REDACTED=REDACTED
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin

HTTP/1.0 200 OK
Date: Tue, 20 Jul 2021 15:05:25 GMT
Server: Apache/2.4.38 (Debian)
Referrer-Policy: no-referrer
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-Robots-Tag: none
X-XSS-Protection: 1; mode=block
X-Powered-By: PHP/7.4.21
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Content-Security-Policy: default-src 'none';base-uri 'none';manifest-src 'self';frame-ancestors 'none'
Content-Encoding: gzip
Feature-Policy: autoplay 'none';camera 'none';fullscreen 'none';geolocation 'none';microphone 'none';payment 'none'
Content-Length: 150
Connection: close
Content-Type: application/json; charset=utf-8

{"req":{"version":"U2F_V2","challenge":"REDACTED","appId":"http:\/\/cloud.dragon-bytes.com"},"sigs":[]}
```

##### Other Thoughts  
Prior to updating the docker container, the YubiKey was working.  I then migrated through the 21.0.3 tagged image and to the latest image, which seems equivalent to the 22.0.0 tag.  After the upgrade, U2F stopped working.  I removed the existing YubiKey and tried adding it back and began seeing this error.  

This seems similar to issue #725.  As I am seeing the same case of the `appId` coming through with an `http` instead of an `https`.  I do have the `"overwriteprotocol": "https"` directive in my `config.php`.  One thing which struck me as odd, is the escaping of the forward slashes in the `appId` field in the pcap I took.  While the escaped forward slash _should_ be equivalent in JavaScript, is this possibly not true in Python  and so the rewrite rule is not being applied correctly?

**Comment by Sylver-Dragon** [comment]
I just managed to find a workaround; however, it is certainly not a fix.  
On my running instance, I edited the file:  
`/var/www/html/custom_apps/twofactor_u2f/lib/Service/U2FManager.php`  
In the function definition for `getU2f` I edited it as seen below:  
```
 private function getU2f(): U2F {
     /** $url = $this->request->getServerProtocol() . '://' . $this->request->getServerHost(); */
     url =  'https://' . $this->request->getServerHost();
     return new U2F($url);
```

I hard coded the function to use `https`, instead of pulling the protocol from the server.

**Comment by Sylver-Dragon** [comment]
To update the workaround to more of a fix, use:  
```
 private function getU2f(): U2F {
                $url = $this->request->getHttpProtocol() . '://' . $this->request->getServerHost();
                return new U2F($url);
```

This change was based on the code in the `OCP/IRequest` interface.  According to the comments for `getHttpProtocol()`: "Returns the used HTTP protocol".  Since the server is behind a Reverse Proxy, this seem to be a better choice to ensure that the correct protocol is used.

**Comment by Sylver-Dragon** [comment]
Ok, so while the above worked for me, more research shows that it shouldn't and I would recommend against doing it.  That said, rolling back the change and trying to get the problem to occur again has resulted in the problem no longer occurring.  In the interest of trying to prevent the problem from coming back, I went ahead and made one further change to my config.  In my nginx config I had the line:  
`proxy_set_header        X-Forwarded-proto $scheme;`  
Looking at the code for `getServerProtocol` in:  
https://github.com/nextcloud/server/blob/5da42273805160f75436f53b829b17e7b4ed2054/lib/private/AppFramework/Http/Request.php#L678
I noticed the case of `HTTP_X_FORWARDED_PROTO`.  While I can't tell if this should matter, I went ahead and changed the nginx config line to:  
`proxy_set_header        X-FORWARDED-PROTO $scheme;`

I can't tell if this did anything, as I can no longer recreate the original problem; but, in case anyone finds this thread it's something to try.
